hack/generator/pkg/astmodel/type_walker.go

```diff
@@ -0,0 +1,195 @@
+
+	state                   typeWalkerState
+	originalVisitTypeName   func(this *TypeVisitor, it TypeName, ctx interface{}) (Type, error)
+	originalVisitObjectType func(this *TypeVisitor, it *ObjectType, ctx interface{}) (Type, error)
+}
+
+type typeWalkerState struct {
+	result     Types
+	processing map[TypeName]struct{}
+}
+
+// NewTypeWalker returns a TypeWalker.
+// The provided visitor VisitTypeName function must return a TypeName and VisitObjectType must return an ObjectType or calls to
+// Walk will panic.
+func NewTypeWalker(allTypes Types, visitor TypeVisitor) *TypeWalker {
+	typeWalker := TypeWalker{
+		allTypes: allTypes,
+	}
+	typeWalker.originalVisitTypeName = visitor.VisitTypeName
+	typeWalker.originalVisitObjectType = visitor.VisitObjectType
+
+	// visitor is a copy - modifications won't impact passed visitor
+	visitor.VisitTypeName = typeWalker.visitTypeName
+	visitor.VisitObjectType = typeWalker.visitObjectType
+
+	typeWalker.visitor = visitor
+	typeWalker.AfterVisitFunc = DefaultAfterVisit
+	typeWalker.MakeContextFunc = IdentityMakeContext
+	typeWalker.WalkCycle = IdentityWalkCycle
+
+	return &typeWalker
+}
+
+func (t *TypeWalker) visitTypeName(this *TypeVisitor, it TypeName, ctx interface{}) (Type, error) {
+	updatedCtx, err := t.MakeContextFunc(it, ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	visitedTypeName, err := t.originalVisitTypeName(this, it, updatedCtx)
+	if err != nil {
+		return nil, err
+	}
+	var ok bool
+	it, ok = visitedTypeName.(TypeName)
+	if !ok {
+		panic(fmt.Sprintf("TypeWalker visitor VisitTypeName must return a TypeName, instead returned %T", visitedTypeName))
+	}
+
+	def, ok := t.allTypes[it]
+	if !ok {
+		return nil, errors.Errorf("couldn't find type %q", it)
+	}
+
+	// Prevent loops by bypassing this type if it's currently being processed. The processing
+	// slice is basically the "path" taken to get to the current type.
+	if _, ok := t.state.processing[def.Name()]; ok {
+		return t.WalkCycle(def, updatedCtx)
+	}
+	t.state.processing[def.Name()] = struct{}{}
+
+	updatedType, err := this.Visit(def.Type(), updatedCtx)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error visiting type %q", def.Name())
+	}
+	updatedDef, err := t.AfterVisitFunc(def, def.WithType(updatedType), updatedCtx)
+	if err != nil {
+		return nil, err
+	}
+
+	delete(t.state.processing, def.Name())
+
+	err = t.state.result.AddWithEqualityCheck(updatedDef)
+	if err != nil {
+		return nil, err
+	}
+
+	return updatedDef.Name(), nil
+}
+
+func (t *TypeWalker) visitObjectType(this *TypeVisitor, it *ObjectType, ctx interface{}) (Type, error) {
+	result, err := t.originalVisitObjectType(this, it, ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	ot, ok := result.(*ObjectType)
+	if !ok {
+		panic(fmt.Sprintf("TypeWalker visitor VisitObjectType must return a *ObjectType, instead returned %T", result))
+	}
+
+	for _, prop := range ot.Properties() {
+		isCycle := isRemoveType(prop.PropertyType())
```

Comment:
Presumably a visitor could return a property with the remove type for other reasons than it containing a cycle? Although I guess that would be a weird thing to do, when the visitor could just remove the property itself in that case rather than needing to ask the walker to do it. The only time it needs to get the walker to do it is if the walker detects a cycle, is that right? And not all of the time, since in that case there'd be no need for a callback. I feel like I need to see the usage to get a feel for how that decision gets made.

Is the WalkCycle interface too flexible? What about just making it a predicate that tells the walker whether to remove the property? I guess that doesn't work since it's in the wrong place. It feels like the interaction between the visitor and the walker is very complicated (but that might be because I can only see half of it).

Reply:
The tests show some sample usage. I do agree that the interplay seems complicated - I think I was struggling to express that in my initial comment on this PR.

The problem with the visitor removing types/properties is that depending on which handler in the visitor makes the decision to remove, you have a nesting problem...
For example let's say that you'd like to remove all `ObjectType` properties that have a particular `Flag` on the `Type`. You have only one way to do that with the visitor as it is today, which is to put the per-property detection logic into `VisitObjectType`, so you'd do something like:

```
func VisitObjectType(...) (Type, error) {
    // loop over all properties
    // check if somewhere in property type hierarchy there is a flag type
    // Check if that flag type has the flag we're looking for, if so, remove.
}
```

That's fine and it works but the issue is that it requires you to "pull up" the processing that could otherwise be done in the `VisitType` for the property. (imagine if instead of removal we just wanted to change the name of every Flag on a FlagType from `Foo` to `Bar` -- we would be doing that in the `VisitFlagType` bit of the visitor). 
In the `TypeWalker` context this "pulling up" is problematic because the function where we want the removal logic to reside is `VisitTypeName` (since that's what's walking to the next type in the graph and where we can do cycle detection easily) - _but_ you can't remove a type in `VisitTypeName`, so instead we would have to pull it up to `VisitObjectType` and do something like:

```
func VisitObjectType(...) (Type, error) {
    // loop over all properties
    // check if one of the has a typename that would trigger a cycle, probably by doing basically what my `VisitTypeName` is doing today
    // Remove the property
    // Call identityVisitObjectType which will eventually call `VisitTypeName` again on all of the existing properties to actually do the walk.
}
```

So it's not impossible to avoid the need for the `TypeWalkerRemoveType` entirely, it just means that the call-graph of what the visitor is doing gets a bit more complicated because there end up being 2 entrypoints to `VisitTypeName` - one that's a pre-check for removal, and then the real one for when we know we don't need to remove.

*Note: In actuality it'd probably be `VisitTypeName` calling some worker function and that worker function being reused in a slight different context (with different args?) in `VisitObjectType`, since as discussed above `VisitTypeName` can't actually convey the concept of removal with its signature today (at least not without some sentinel `RemovalType` like I've added.

Maybe it would make sense to promote this removal capability into visitor itself somehow? (That does just move complexity around a bit but might be more natural)

Reply:
To your original comment though, yes the variable here shouldn't be `isCycle`, since the idea was you can remove for reasons other than cycles.

Reply:
I've pushed a draft PR that shows actual usage here: https://github.com/Azure/k8s-infra/pull/422/files#diff-cd9aa548dd1752cd7d7998f99bc00ef5c2e6e69154d26efc1e8e02e376cb9de3R268

This PR looks bigger than it really is because it includes this change and a bunch of golden files tests, so don't be too afraid of it. Don't bother reviewing it yet though, I'll mark it as not draft once it's ready for a real review.